### PR TITLE
Move index setting outside of expansion

### PR DESCRIPTION
**Pull request by harsh-nod**
This PR separates the setting of indices pre-
and post-expansion from expansion itself. This
separation of concerns should make the overall
design more modular.

### CODE REVIEW

File: iree/turbine/kernel/wave/index_sequence_analysis.py [hunk 1]
```diff
@@ -94,3 +100,159 @@ def has_strided_access(node: fx.Node) -> bool:
...
+    """
+    Add offsets to the indices based on the expanded dims.
+    """
+    hw_cons = get_hardware_constraint(constraints)
+
+    def apply_offset(node: fx.Node):
+        custom = get_custom(node)
+        if custom.expanded_dims is None:
+            return False
+        for dim, scale in custom.expanded_dims.items():
```
[the remark this answers is not in the document]
Comment (raikonenfnu) [reply iree/turbine/kernel/wave/index_sequence_analysis.py:253]: I guess scale would have to be the "expansion idx"/iteration on dim scaling?

[the remark this answers is not in the document]
Comment (raikonenfnu) [reply iree/turbine/kernel/wave/index_sequence_analysis.py:253]: actually, yeah this seems to be "restricted" `dim_query`, thanks! :)

[the remark this answers is not in the document]
Comment (harsh-nod) [reply iree/turbine/kernel/wave/index_sequence_analysis.py:253]: Yes this is a dictionary that specifies the dimension and dim_scaling so something like
```
{M : 0, N: 1}
```
By bringing this out, we can now play with the expansion implementation and the only thing the expansion implementation has to satisfy is that it needs to annotate each node with this expanded dims dictionary.

[the remark this answers is not in the document]
Comment (harsh-nod) [reply iree/turbine/kernel/wave/index_sequence_analysis.py:253]: Yes you are right, it is "restricted".

File: iree/turbine/kernel/wave/index_sequence_analysis.py [hunk 2]
```diff
@@ -94,3 +100,159 @@ def has_strided_access(node: fx.Node) -> bool:
...
+            if dim in custom.index:
+                custom.index[dim].start += scale * hw_cons.vector_shapes[dim]
```
Comment (raikonenfnu) [inline iree/turbine/kernel/wave/index_sequence_analysis.py:255]: What if the dim is in mma_indices? Won't we need 
```
vector_size = hardware_constraints[0].mma_matrix_shapes[
                    mma_indices[constraint.dim]
                ]
```

File: iree/turbine/kernel/ops/wave_ops.py [hunk 3]
```diff
@@ -498,6 +498,18 @@ def scheduling_parameters(self, value: Any):
             raise ValueError("Scheduling parameters must be a dict")
         self.fx_node.scheduling_parameters = value
 
+    @property
+    def expanded_dims(self) -> dict[IndexSymbol, int]:
```
Comment (raikonenfnu) [inline iree/turbine/kernel/ops/wave_ops.py:502]: Can we add documentation that mentions that it is used to store it's "expansion id"/"dim_query"?

File: iree/turbine/kernel/wave/index_sequence_analysis.py (hunk 2 above)
Comment (raikonenfnu) [reply iree/turbine/kernel/wave/index_sequence_analysis.py:255]: Resolved offline, vector_size and mma_size is now fused inside https://github.com/iree-org/iree-turbine/blob/77b60ef44670bcef372a00fa578ace3c2fa60216/iree/turbine/kernel/wave/utils.py#L225-L231

File: iree/turbine/kernel/ops/wave_ops.py (hunk 3 above)
Comment (harsh-nod) [reply iree/turbine/kernel/ops/wave_ops.py:502]: added.

**Review by raikonenfnu — APPROVED** [review]

**Pull request merged**